### PR TITLE
Distinct for non-indexed and linked properties

### LINKED ISSUE #1906: Add support for distinct for links
i tried 

```
realm.distinct(Parent.class, "Child");
```

throw `java.lang.UnsupportedOperationException: The field must be indexed before distinct() can be used.`
but linked field can not be indexed

and tried

```
realm.distinct(Parent.class, "Child.id");
```

throw `java.lang.IllegalArgumentException: Field name 'Child.id' does not exist.`

compile 'io.realm:realm-android:0.86.0'
android v4.4.4

### PR DESCRIPTION
**Pull request by rorbech**
Lifting restriction on `distinct` on non-indexed and linked properties according to support in Core. 

Distinct is now support on all field types except lists. 

Closes #1906

### CODE REVIEW

File: realm/realm-library/src/androidTest/java/io/realm/RealmQueryTests.java [hunk 1]
```diff
@@ -3705,5 +3867,80 @@ public void limit_invalidValuesThrows() {
...
+                if (!object.isNull(fieldName)) {
+                    Object e = object.get(fieldName);
+                    // Need to convert byte arrays to list to detect duplicates when inserting to values
+                    if (e instanceof byte[]) {
+                        elements.add(convertBytesToList((byte[]) e));
+                    } else {
+                        elements.add(e);
+                    }
+                } else {
+                    elements.add(NULL_PLACEHOLDER);
```
Comment (cmelchior) [inline realm/realm-library/src/androidTest/java/io/realm/RealmQueryTests.java:3930]: Not sure I understand this?

File: realm/realm-library/src/androidTest/java/io/realm/RealmQueryTests.java [hunk 2]
```diff
@@ -3344,46 +3423,104 @@ public void distinct_async_doesNotExist() {
...
+        final long numberOfBlocks = 3;
+        final long numberOfObjects = 3;
+        final Realm realm = looperThread.getRealm();
+        populateForDistinct(realm, numberOfBlocks, numberOfObjects, false);
+
+        DynamicRealm dynamicRealm = DynamicRealm.createInstance(realm.sharedRealm);
+        RealmResults<DynamicRealmObject> all = dynamicRealm.where(AnnotationIndexTypes.CLASS_NAME)
+                .findAll();
+
+        for (String field : AnnotationIndexTypes.NOT_INDEX_FIELDS) {
```
Comment (cmelchior) [inline realm/realm-library/src/androidTest/java/io/realm/RealmQueryTests.java:3497]: Does this indexing distincting still makes sense or should it be combined with the above test case?

File: realm/realm-library/src/androidTest/java/io/realm/RealmQueryTests.java [hunk 3]
```diff
@@ -59,6 +66,8 @@
 @RunWith(AndroidJUnit4.class)
 public class RealmQueryTests extends QueryTests {
 
+    private static final Object NULL_PLACEHOLDER = new Object();
```
[the remark this answers is not in the document]
Comment (rorbech) [reply realm/realm-library/src/androidTest/java/io/realm/RealmQueryTests.java]: Eliminated after updating manual distinct implementation as new implementation supports adding null values directly.

File: realm/realm-library/src/androidTest/java/io/realm/RealmQueryTests.java [hunk 4]
```diff
@@ -3120,95 +3151,143 @@ public void distinct_failIfAppliedMultipleTimes() {
                 .distinct(AnnotationIndexTypes.FIELD_INDEX_DATE);
     }
 
-    @Test
-    public void distinct_notIndexedFields() {
+    private void distinctAllFields(Realm realm, String prefix) {
```
[the remark this answers is not in the document]
Comment (rorbech) [reply realm/realm-library/src/androidTest/java/io/realm/RealmQueryTests.java:3154]: Added a number of comments to assist in reasoning.

File: realm/realm-library/src/androidTest/java/io/realm/RealmQueryTests.java (hunk 1 above)
Comment (rorbech) [reply realm/realm-library/src/androidTest/java/io/realm/RealmQueryTests.java]: Keeping track of null values as they are still considered distinct. Previous implementation used container not allow null values directly. The current list allows null values, so eliminated the `NULL_PLACEHOLDER`.

File: realm/realm-library/src/androidTest/java/io/realm/RealmQueryTests.java (hunk 2 above)
Comment (rorbech) [reply realm/realm-library/src/androidTest/java/io/realm/RealmQueryTests.java]: As per discussion I have removed explicit differentiation of testing `distinct` on both indexed and non-indexed fields. The synchronous tests covers all types for direct, linked and nested linked objects. Async and multi argument distinct are only smoke tested for some types and unsupported types.

**Pull request merged**